### DISCUSSION TITLE
Do you have any concerns about the development prospects of OpenCl?

### DISCUSSION THREAD

**Discussion by wujilingfeng** [diku-dk/futhark#1944] [Q&A] [answered]
I see that this library has excellent support for opencl. But according to the information I have received, opencl will integrate into Vulkan, and the ecology of opencl is also not good. I would like to know your opinion on this.

**Comment by athas** [comment]
We want to add a Vulkan backend eventually, but it's not immediately planned.  We made a prototype once that worked OK, but we encountered some friction because Vulkan compute is/was not as flexible as we need.

We also have a CUDA backend that is fully operational.

**Comment by rowanG077** [comment, marked as answer]
Personally I would love if openCL stays, at least in the short to medium term. I deploy my code to a machine with an NVIDIA GPU. But in my local development I only have an integrated Intel GPU. The difference between openCL and CUDA wasn't that great so I still use openCL on my production machine.

Another aspect is that I would like to move to an Asahi Linux powered Macbook. Which soon should support openCL but Vulkan is still pretty far away. So unless Vulkan compute offers tangible benefits over openCL I would rather like if Futhark keeps supporting openCL.

**Comment by athas** [comment]
We certainly have no plans for removing or deprecating the OpenCL backend.  It is the main API supported by AMD (I think; AMD is not great at communication).  If anything, we want to make the GPU backends more generic so we can easily target other APIs (Vulkan, WebGPU) as well.  The host level concepts should be very similar, so the main challenge is generating the kernel code - but that is also not very difficult.

**Reply by wujilingfeng** [reply]
In China, the information I see about OpenCl is mostly negative, so I also feel anxious. But the environment in China is not the same as abroad, for example, there are almost no people learning functional languages in China. So I want to hear more professional and diverse perspectives. You have done excellent work, come on!
